### ISSUE TITLE
Different behavior of FromKeyedServices attribute in Minimal APIs handler vs class library

### ISSUE THREAD

**Issue by marcominerva** [dotnet/aspnetcore#55545]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

If we use the `FromKeyedServices` attribute in a class library and the keyed registration is missing, it falls back to the default dependency injection behavior and gets the last registered implementation. However, If we use `FromKeyedServices` directly in a Minimal API endpoint handler and the key does not exist, it throws an exception.

### Expected Behavior

The FromKeyedServices should behave in the same way in every context.

### Steps To Reproduce

Minimal repro here: https://github.com/marcominerva/KeyedServicesIssue

### .NET Version

8.0.204

**Comment by captainsafia** [comment]
@marcominerva Thanks for reporting this!

The behavior that you're seeing is a consistent and intentional choice with all the implementations for keyed services in ASP.NET Core (SignalR, MVC, minimal APIs).

The expected pattern for resolving a service instance with any key using keyed DI is to take advantage of the `KeyedService.AnyKey` definition which is designed to match the behavior that you see here.

cc: @benjaminpetit For insights onto the behavior in the runtime's DI resolution strategy

**Comment by marcominerva** [comment]
So, to be sure to have understood the context, in the following case:

```csharp
builder.Services.AddSingleton<SampleConfig>();

app.MapGet("api/sample-config", ([FromKeyedServices("x")] SampleConfig config) =>
{
    return TypedResults.Ok(config.Value);
});
```

This code throws an exception, because there isn't a `SampleConifg` instance that has been registered with the key **x**.

On the other hand, in the following scenario:

```csharp
builder.Services.AddSingleton<SampleConfig>();
builder.Services.AddScoped<OtherSampleService>();

app.MapGet("api/external-sample-config", (OtherSampleService otherSampleService) =>
{
    return TypedResults.Ok(otherSampleService.ConfigValue);
})

public class OtherSampleService
{
    private readonly SampleConfig config;

    public OtherSampleService([FromKeyedServices("x")] SampleConfig config, IServiceProvider serviceProvider)
    {
        this.config = config;
    }

    public string ConfigValue => config.Value.ToString();
}
```

The code doesn't raise any exception and, even if `OtherSampleService` requires a `SampleConfig` with key **x**, because it doesn't exist, it gets the "default" registration.

**Comment by captainsafia** [comment]
> So, to be sure to have understood the context, in the following case:

Yep, your understanding here is correct. I agree with you that the difference in behavior is a little strange so I've pinged @benjaminpetit above (who worked on the keyed services implementation) to get a sense of why the implementation behaves this way in constructor binding.
